[FEATURE] API for editing of raster data

wonder-sk opened the pull request:
Until now it was not possible to edit raster layer data because rasters were always opened in read-only mode (only newly created providers with QgsRasterDataProvider::create() would be in read/write mode).

To create a 2x2 raster block with one byte per pixel:

```
block = QgsRasterBlock(Qgis.Byte, 2, 2)
block.setData(b'\xaa\xbb\xcc\xdd')
```

To overwrite existing raster data at position 0,0 by the 2x2 block:

```
provider.setEditable(True)
provider.writeBlock(block, band, 0, 0)
provider.setEditable(False)
```

wonder-sk committed:
[FEATURE] API to enable/disable editing of raster data

To create a 2x2 raster block with one byte per pixel:

```
block = QgsRasterBlock(Qgis.Byte, 2, 2)
block.setData(b'\xaa\xbb\xcc\xdd')
```

To overwrite existing raster data at position 0,0 by the 2x2 block:

```
provider.setEditable(True)
provider.writeBlock(block, band, 0, 0)
provider.setEditable(False)
```

NathanW2 commented:
Holy heck yeah!

Now we just need a paint brush tool and we can remake GIMP ;)

On Thu, Jan 19, 2017 at 2:06 PM, Martin Dobias <notifications@github.com>
wrote:

> Until now it was not possible to edit raster layer data because rasters
> were always opened in read-only mode (only newly created providers with
> QgsRasterDataProvider::create() would be in read/write mode).
>
> To create a 2x2 raster block with one byte per pixel:
>
> block = QgsRasterBlock(Qgis.Byte, 2, 2)
> block.setData(b'\xaa\xbb\xcc\xdd')
>
> To overwrite existing raster data at position 0,0 by the 2x2 block:
>
> provider.setEditable(True)
> provider.writeBlock(block, band, 0, 0)
> provider.setEditable(False)
>
> ------------------------------
> You can view, comment on, or merge this pull request online at:
>
>   https://github.com/qgis/QGIS/pull/4022
> Commit Summary
>
>    - [FEATURE] API to enable/disable editing of raster data
>
> File Changes
>
>    - *M* python/core/raster/qgsrasterblock.sip
>    <https://github.com/qgis/QGIS/pull/4022/files#diff-0> (10)
>    - *M* python/core/raster/qgsrasterdataprovider.sip
>    <https://github.com/qgis/QGIS/pull/4022/files#diff-1> (34)
>    - *M* src/core/raster/qgsrasterblock.cpp
>    <https://github.com/qgis/QGIS/pull/4022/files#diff-2> (17)
>    - *M* src/core/raster/qgsrasterblock.h
>    <https://github.com/qgis/QGIS/pull/4022/files#diff-3> (10)
>    - *M* src/core/raster/qgsrasterdataprovider.cpp
>    <https://github.com/qgis/QGIS/pull/4022/files#diff-4> (12)
>    - *M* src/core/raster/qgsrasterdataprovider.h
>    <https://github.com/qgis/QGIS/pull/4022/files#diff-5> (34)
>    - *M* src/providers/gdal/qgsgdalprovider.cpp
>    <https://github.com/qgis/QGIS/pull/4022/files#diff-6> (34)
>    - *M* src/providers/gdal/qgsgdalprovider.h
>    <https://github.com/qgis/QGIS/pull/4022/files#diff-7> (4)
>    - *M* tests/src/core/testqgsrasterblock.cpp
>    <https://github.com/qgis/QGIS/pull/4022/files#diff-8> (91)
>
> Patch Links:
>
>    - https://github.com/qgis/QGIS/pull/4022.patch
>    - https://github.com/qgis/QGIS/pull/4022.diff
>
> —
> You are receiving this because you are subscribed to this thread.
> Reply to this email directly, view it on GitHub
> <https://github.com/qgis/QGIS/pull/4022>, or mute the thread
> <https://github.com/notifications/unsubscribe-auth/AAXS3M-z44o7D-H0UWNR2mL1HBn3F44Vks5rTuE0gaJpZM4Lnrp7>
> .
>

NathanW2 commented:
This could also mean our raster calculator could do live updating of a
raster layer right?


On Thu, Jan 19, 2017 at 2:07 PM, Nathan Woodrow <madmanwoo@gmail.com> wrote:

> Holy heck yeah!
>
> Now we just need a paint brush tool and we can remake GIMP ;)
>
> On Thu, Jan 19, 2017 at 2:06 PM, Martin Dobias <notifications@github.com>
> wrote:
>
>> Until now it was not possible to edit raster layer data because rasters
>> were always opened in read-only mode (only newly created providers with
>> QgsRasterDataProvider::create() would be in read/write mode).
>>
>> To create a 2x2 raster block with one byte per pixel:
>>
>> block = QgsRasterBlock(Qgis.Byte, 2, 2)
>> block.setData(b'\xaa\xbb\xcc\xdd')
>>
>> To overwrite existing raster data at position 0,0 by the 2x2 block:
>>
>> provider.setEditable(True)
>> provider.writeBlock(block, band, 0, 0)
>> provider.setEditable(False)
>>
>> ------------------------------
>> You can view, comment on, or merge this pull request online at:
>>
>>   https://github.com/qgis/QGIS/pull/4022
>> Commit Summary
>>
>>    - [FEATURE] API to enable/disable editing of raster data
>>
>> File Changes
>>
>>    - *M* python/core/raster/qgsrasterblock.sip
>>    <https://github.com/qgis/QGIS/pull/4022/files#diff-0> (10)
>>    - *M* python/core/raster/qgsrasterdataprovider.sip
>>    <https://github.com/qgis/QGIS/pull/4022/files#diff-1> (34)
>>    - *M* src/core/raster/qgsrasterblock.cpp
>>    <https://github.com/qgis/QGIS/pull/4022/files#diff-2> (17)
>>    - *M* src/core/raster/qgsrasterblock.h
>>    <https://github.com/qgis/QGIS/pull/4022/files#diff-3> (10)
>>    - *M* src/core/raster/qgsrasterdataprovider.cpp
>>    <https://github.com/qgis/QGIS/pull/4022/files#diff-4> (12)
>>    - *M* src/core/raster/qgsrasterdataprovider.h
>>    <https://github.com/qgis/QGIS/pull/4022/files#diff-5> (34)
>>    - *M* src/providers/gdal/qgsgdalprovider.cpp
>>    <https://github.com/qgis/QGIS/pull/4022/files#diff-6> (34)
>>    - *M* src/providers/gdal/qgsgdalprovider.h
>>    <https://github.com/qgis/QGIS/pull/4022/files#diff-7> (4)
>>    - *M* tests/src/core/testqgsrasterblock.cpp
>>    <https://github.com/qgis/QGIS/pull/4022/files#diff-8> (91)
>>
>> Patch Links:
>>
>>    - https://github.com/qgis/QGIS/pull/4022.patch
>>    - https://github.com/qgis/QGIS/pull/4022.diff
>>
>> —
>> You are receiving this because you are subscribed to this thread.
>> Reply to this email directly, view it on GitHub
>> <https://github.com/qgis/QGIS/pull/4022>, or mute the thread
>> <https://github.com/notifications/unsubscribe-auth/AAXS3M-z44o7D-H0UWNR2mL1HBn3F44Vks5rTuE0gaJpZM4Lnrp7>
>> .
>>
>
>

wonder-sk commented:
> This could also mean our raster calculator could do live updating of a raster layer right?

Indeed :-)

NathanW2 commented:
![kool-aid](https://cloud.githubusercontent.com/assets/381660/22093504/f8d28384-de51-11e6-8f00-586e1ab5597f.jpg)

nirvn commented:
@wonder-sk , I'd love for this to bring us one step closer to being able to attach virtual raster bands to a raster. Ie an expression that combines data from other bands into a virtual one.

nyalldawson commented on python/core/raster/qgsrasterdataprovider.sip:
```diff
@@ -233,10 +233,44 @@ class QgsRasterDataProvider : QgsDataProvider, QgsRasterInterface
+     * @see isEditable(), writeBlock()
+     * @note added in QGIS 3.0
+     */
+    virtual void setEditable( bool enabled );
```
Is it worth making this return a bool to indicate whether the switch worked?

wonder-sk replied:
Yeah it would probably make sense... I wanted to keep the setter "clean" with no return value, but here a return value may be useful :-)

wonder-sk commented:
@nirvn I think in general this kind of functionality would be better implemented using our raster pipe. We would just need a new QgsRasterInterface implementation that would transform band's data (according to raster expression) and pass the transformed block further down the pipeline...

NathanW2 commented:
Maybe just a string so you can return an error message?   Having a bool
with no reason is painful at times.

On Thu, Jan 19, 2017 at 2:33 PM, Martin Dobias <notifications@github.com>
wrote:

> *@wonder-sk* commented on this pull request.
> ------------------------------
>
> In python/core/raster/qgsrasterdataprovider.sip
> <https://github.com/qgis/QGIS/pull/4022>:
>
> > +    /** Checks whether the provider is in editing mode, i.e. raster write operations will be accepted.
> +     * By default providers are not editable. Use setEditable() method to enable/disable editing.
> +     * @see setEditable(), writeBlock()
> +     * @note added in QGIS 3.0
> +     */
> +    virtual bool isEditable() const;
> +
> +    /** Turns on/off editing mode of the provider. When in editing mode, it is possible
> +     * to overwrite data of the provider using writeBlock() calls.
> +     * @note Only some providers support editing mode and even those may fail to turn turn
> +     * the underlying data source into editing mode, so it is necessery to check afterwards
> +     * with isEditable() whether the operation was successful.
> +     * @see isEditable(), writeBlock()
> +     * @note added in QGIS 3.0
> +     */
> +    virtual void setEditable( bool enabled );
>
> Yeah it would probably make sense... I wanted to keep the setter "clean"
> with no return value, but here a return value may be useful :-)
>
> —
> You are receiving this because you commented.
> Reply to this email directly, view it on GitHub
> <https://github.com/qgis/QGIS/pull/4022>, or mute the thread
> <https://github.com/notifications/unsubscribe-auth/AAXS3Gy7ptT-1XVJjK2dLtmku1dK78Ufks5rTueJgaJpZM4Lnrp7>
> .
>

wonder-sk committed:
Make setEditable() return true/false to indicate success/error

wonder-sk committed:
Convenience API call to create empty raster file

wonder-sk committed:
Make the one band raster creation method return data provider

wonder-sk commented:
Added one more API call to simplify creation of raster layers:
```
writer = QgsRasterFileWriter("/tmp/my_file.tif")
writer.createOneBandRaster(Qgis.Byte, 100, 100, extent, crs)
```
Also changed setEditable() to return bool (sorry no error msg...)

nyalldawson commented:
> Maybe just a string so you can return an error message?   Having a bool
with no reason is painful at times.

Just seems odd to me to have to do:

    layer.setEditable(True)
    if layer.editable():
      ...

vs

    if layer.setEditable(True):

nyalldawson commented:
> writer = QgsRasterFileWriter("/tmp/my_file.tif")
> provider = writer.createOneBandRaster(Qgis.Byte, 100, 100, extent, crs)

@wonder-sk can you do something like this for vector layers please? ;) It's a pita to create vector layers from scratch in current api...

NathanW2 commented:
I don't really mind what the API is as long as I can get a reason.

if layer.setEditable(True):
 ....
else:
 layer.lastErrors()





On Thu, Jan 19, 2017 at 3:46 PM, Nyall Dawson <notifications@github.com>
wrote:

> writer = QgsRasterFileWriter("/tmp/my_file.tif")
> provider = writer.createOneBandRaster(Qgis.Byte, 100, 100, extent, crs)
>
> @wonder-sk <https://github.com/wonder-sk> can you do something like this
> for vector layers please? ;) It's a pita to create vector layers from
> scratch in current api...
>
> —
> You are receiving this because you commented.
> Reply to this email directly, view it on GitHub
> <https://github.com/qgis/QGIS/pull/4022#issuecomment-273687742>, or mute
> the thread
> <https://github.com/notifications/unsubscribe-auth/AAXS3LTOboN4YUhJ-7C5VzY0IY4l0gRjks5rTvi_gaJpZM4Lnrp7>
> .
>

NathanW2 commented:
I don't really mind the API as long as I can get the error:

```
if layer.startEditing(True)
else:
   print layer.errors()
````

wonder-sk commented:
@NathanW2 for GDAL the most likely cause of error is reported through ```QgsDataProvider::error()``` ...

@nyalldawson yeah such thing for vectors would be useful too... :-)
